protoc-gen-kotlin/protoc-gen-kotlin-common/src/main/kotlin/pbandk/gen/CodeGenerator.kt

```diff
@@ -258,6 +301,126 @@ open class CodeGenerator(val file: File, val kotlinTypeMappings: Map<String, Str
+            line("$fullTypeName.JsonMapper(").indented {
+                val allFields = type.sortedStandardFieldsWithOneOfs()
+                allFields.forEachIndexed { i, (field, _) ->
+                    when {
+                        (field.type == File.Field.Type.ENUM) -> {
+                            when {
+                                field.repeated -> lineBegin("${field.kotlinFieldName}.mapNotNull { it.name }")
+                                else -> lineBegin("${field.kotlinFieldName}?.name")
+                            }
+                        }
+                        (field.type == File.Field.Type.MESSAGE) -> {
+                            when {
+                                field.map -> {
+                                    when {
+                                        field.mapEntry()!!.mapEntryValueIsMessage -> lineBegin("${field.kotlinFieldName}.mapValues { it.value?.toJsonMapper() }")
+                                        field.mapEntry()!!.mapEntryValueIsEnum -> lineBegin("${field.kotlinFieldName}.mapValues { it.value?.name }")
+                                        else -> lineBegin(field.kotlinFieldName)
+                                    }
+                                }
+                                field.repeated -> lineBegin("${field.kotlinFieldName}.map { it.toJsonMapper() }")
+                                else -> lineBegin("${field.kotlinFieldName}?.toJsonMapper()")
+                            }
+                        }
+                        else -> lineBegin(field.kotlinFieldName)
+                    }
+                    lineEnd(if (i == allFields.lastIndex) "" else ",")
+                }
+            }.line(")")
+        }
+    }
+
+    protected fun writeJsonMapperToMessageExtension(type: File.Type.Message, fullTypeName: String) {
+        line().line("fun $fullTypeName.JsonMapper.toMessage(): $fullTypeName =").indented {
+            line("$fullTypeName(").indented {
+                type.fields.forEachIndexed { i, field ->
+                    when (field) {
+                        is File.Field.OneOf -> {
+                            line("${field.kotlinFieldName} = ").indented {
+                                field.fields.forEachIndexed { j, oneOfField ->
+                                    val prefix = if (j == 0) "" else " ?: "
+                                    val suffix = if (j == field.fields.lastIndex && i != type.fields.lastIndex) "," else ""
+                                    line(prefix +
+                                            "${oneOfField.kotlinFieldName}?.let { value -> " +
+                                            "$fullTypeName.${field.kotlinTypeName}.${field.kotlinFieldTypeNames[oneOfField.name]}(" +
+                                            when {
+                                                (oneOfField.type == File.Field.Type.ENUM) -> {
```

Comment:
🙃 Change to `when (oneOfField.type) { ... }`

Reply:
👍